### PR TITLE
Set prometheus' data PV's reclaimPolicy to Retain

### PR DESCRIPTION
**Pull request by yuvipanda**
Helps us recover prometheus data in case of accidental deletion

Fixes https://github.com/2i2c-org/infrastructure/issues/2717

### CODE REVIEW

**Review by GeorgianaElena — APPROVED** [review]

**Comment by sgibson91** [comment]
Is this Python code a workaround to edit our currently deployed clusters? Is this retention policy something we can set as default in the support chart values?

I'm worried this PR is introducing more "deployer overwrites helm stuff that makes R2R more difficult" and so need help understanding the context.

**Comment by consideRatio** [comment]
I'd like to avoid introducing this unless we are confident we need this patch permanently, and if we do convey that via a comment. This requires further investigation so here goes:

[From these k8s docs](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#delete), we learn that...

> (Persistent) Volumes that were dynamically provisioned inherit the [reclaim policy of their StorageClass](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#reclaim-policy), which defaults to Delete. The administrator should configure the StorageClass according to users' expectations; otherwise, the PV must be edited or patched after it is created. See [Change the Reclaim Policy of a PersistentVolume](https://kubernetes.io/docs/tasks/administer-cluster/change-pv-reclaim-policy/).

Related k8s api references:
- [PVC](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#persistentvolumeclaim-v1-core)
- [PV](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#persistentvolume-v1-core)
- [StorageClass](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#storageclass-v1-storage-k8s-io)

It seems clear from looking at the api reference that PVC can't be used to define the reclaim policy on the PV, but PVCs are whats the Helm chart can control with regards to the requesting storage. So, from the chart config, we can at best adjust the StorageClass which can influence the reclaim policy, but then we must manage a StorageClass which is cloud specific etc.

### What to do?

1. Deployer to patch prometheus PV on support chart deploy (this PR)
2. Patching all existing PVs once and...
  - a) Deployer to patch prometheus PV as manually run command, and referencing it from cluster creation docs
  - b) Patching existing clusters default StorageClass resources to have a `reclaimPolicy` of `Retain`, and referencing doing this from cluster creation docs
  - c) Create a dedicated StorageClass in each cluster for the prometheus PVs
    - This is a very bad option I think, because we need to declare StorageClass resources that are cloud provider specific, and we must explicitly declare the storageclass we want to use in the prometheus chart's PVC config for each cluster I think - one default value won't work well.

I'm not so opinionated on any choice above besides not ending up creating our own StorageClass (2c).

**Review by consideRatio — APPROVED** [review]
I made some code suggestions, supposed only to be comments + refactor rather than a change in outcome!

I don't think we have much options unless we want to define Retain on the clusters default storageclass, but I suggest we would consider that separately from this PR if we want to go for it.

File: deployer/cluster.py [hunk 1]
```diff
@@ -34,6 +34,45 @@ def auth(self):
...
+            print("Setting Prometheus Data Volume's ReclaimPolicy to Retain")
+            # Set it to retain only when it is not
+            patch = {"spec": {"persistentVolumeReclaimPolicy": "Retain"}}
+            with tempfile.NamedTemporaryFile() as f:
+                f.write(json.dumps(patch).encode())
+                f.flush()
+
+                subprocess.check_call(
+                    ["kubectl", "patch", "pv", pv_name, "--patch-file", f.name]
+                )
```
[the remark this answers is not in the document]
Comment (yuvipanda) [reply deployer/cluster.py:74]: @consideRatio I just measured doing it with and without the conditional check, and just unconditionally doing the patch (as you suggested) is faster! So I've removed the conditional check.

**Review by consideRatio — APPROVED** [review]
I think this was the strategy we kind of have to go for since: the prometheus chart doesn't manage the PV directory, only though the PVC, and the PV gets the reclaim policy based on the StorageClass, the only viable options seems to patch the StorageClass and/or patch the PV.

**Comment by yuvipanda** [comment]
@sgibson91 thank you for raising the r2r concern. I think I'll try to open another PR codifying our R2R guidelines for what can and can not go in the deployer, and then we can merge this one.

**Comment by yuvipanda** [comment]
Thanks to both @sgibson91's point about R2R as well as performance considerations, I'm going to close this in favor of things like https://github.com/2i2c-org/infrastructure/pull/3112. The storageclasses thus end up in the helm chart, rather than in the deployer.

**Pull request closed without merging**